text/metrics/0118-statsd-and-prometheus-exporters.md

```diff
@@ -0,0 +1,226 @@
+| SumObserver       | No  | Adding   | Yes | Yes | Sum |
+| UpDownSumObserver | No  | Adding   | No  | Yes | Sum |
+| ValueObserver     | No  | Grouping | n/a | n/a | MinMaxLastSumCount |
+
+Note that the Precomputed-Sum property places some constraints on the
+combination of Aggregator and Exporter.  To compute _Delta_
+aggregations of Precomputed-Sum instruments requires an aggregation
+that supports subtraction (which MUST include Sum and SHOULD include
+Histogram).
+
+### Prometheus
+
+This is based off of the Prometheus [Metric
+Types](https://prometheus.io/docs/concepts/metric_types).
+
+#### Prometheus instruments
+
+Prometheus Counter instruments are semantically identical to
+OpenTelemetry Counter instruments, including the Mnotonic property.
+They are exposed as a Cumulative Sum aggregation.
+
+Prometheus Gauge instruments have a number of uses, depending on
+whether they are used as an adding or as a grouping instrument.  They
+are exposed as a single data point equal to the the last value that
+was set.  Because Prometheus clients are stateful, Gauges support both
+`Set()` and `Add()` methods.  Generally, Prometheus Gauges used to
+`Add()` map into OpenTelemetry UpDownCounter instruments (maybe
+SumObserver, UpDownSumObserver), while Prometheus Gauges used to
+`Set()` map into OpenTelemetry ValueRecorder instruments (maybe
+ValueObserver).
+
+Prometheus Histogram instruments are exposed as Cumulative
+aggregations, defined by a number of bucketed counts, accumulated from
+the start of the process.  Prometheus Histogram instruments map into
+ValueRecorder instruments (maybe ValueObserver).  To configure a
+Histogram exposition in the Prometheus exporter, configure a Histogram
+Aggregator with the desired buckets.
+
+Prometheus Summary instruments are discouraged, as mentioned above,
+because they are not mergeable.  Uses of Prometheus Summary
+instruments map into OpenTelemetry ValueRecorder instruments (maybe
+ValueObserver).  To configure a Summary exposition in the Prometheus
+exporter, configure a Sketch Aggregator with the desired quantiles.
+
+#### OpenTelemetry Aggregator to Prometheus exposition
+
+The following table lists the mapping from Aggregator to Prometheus
+exposition format.  The "Typical Instruments" listed are the
+applicable OpenTelemetry instruments, for which the Prometheus mapping
+is sensible.
+
+| OpenTelemetry aggregator | Default Prometheus data type mapping | Export kind | Typical instruments | Notes |
+| -- | -- | -- | -- | -- |
+| Sum (Monotonic)       | Counter    | Cumulative | Counter(*), SumObserver(*) | |
+| Sum (Non-Monotonic)   | Gauge      | Cumulative | UpDownCounter(*), UpDownSumObserver(*) | |
+| LastValue             | Gauge      | Cumulative | ValueRecorder, ValueObserver, SumObserver, UpDownSumObserver | |
+| MinMaxLastSumCount    | Gauge      | Delta      | ValueRecorder(*), ValueObserver(*) | Expose the LastValue field as the Gauge |
```

Comment:
The notes say that LastValue will be exposed as a Gauge.

Is the idea that we will throw away the other non-last value (min/max/sum/count) by default?

I would have thought a more natural default would be for each of these aggregations to be represented as a separate Gauge, otherwise we are aggregating that data just to throw it away - i.e.

```yaml
metricname_min
metricname_max
metricname_last
metricname_sum
metricname_count
```

That would obviously have some implications for the mapping from Prometheus -> OT -> Prometheus that I'm not too sure how to address.

Reply:
What you propose is a viable export strategy, I just don't believe it's the one most Prometheus users want or expect. This is what the Prometheus "Summary" metric exports, approximately. If we replaced "metricname_last" with just "metricname" that would create less confusion for Prometheus users, but they might not appreciate the cost of indexing new metric names that they never wanted.

It would be more optimal, but much more complex, to try to match the default aggregation to the configured exporter. If an OTLP exporter is configured, then the question becomes recursive. What aggregation should I use to satisfy a downstream exporter? That would create a headache at startup.

The idea of adding an optimization (#117) to store a single value when there is in fact a single value, was meant to address some of the concern about throwing away information. I believe some platforms want to see min/max/sum/count and would not like to force users of those systems to reconfigure the aggregation.

Reply:
Makes sense. Thanks for the detailed explanation